src/regmodsm/dimension.py

```diff
@@ -0,0 +1,105 @@
+import itertools
+import numpy as np
+import pandas as pd
+from pandas import DataFrame
+from scipy.sparse import csc_matrix
+
+
+class Dimension:
+    """Dimension used for grouped variable smoothing.
+
+    Parameters
+    ----------
+    name : str or list of str
+        Name of the dimension column(s) in the data. When it is a list
+        of str, it is used for multi-indexing.
+    type : str or list of str
+        Dimension type(s), either "numerical" or "categorical". When it
+        is a list of str, it is used for multi-indexing.
+    label : str, optional
+        Name of dimension in the model.
+
+    """
+
+    def __init__(
+        self,
+        name: str | list[str],
+        type: str | list[str],
+        label: str | None = None,
+    ) -> None:
+        self.name = [name] if isinstance(name, str) else list(name)
+        self.type = [type] if isinstance(type, str) else list(type)
```

Comment:
Do the dimension types need to be the same for multi-indexing? Should we add a check that the list is the same length? etc.

Reply:
In Model, do you need to change how you access dimension types for lists?

Reply:
> Do the dimension types need to be the same for multi-indexing? Should we add a check that the list is the same length? etc.

Thanks very much @kels271828!! I haven't think the design of this part through yet. I think the better way to dealt with this is to ask user to input a list of tuple with `name` and `type` as a pair, so that user is likely to make mistake. However this will break the backward compatibility, I will try to address this in another PR later on :)



> In Model, do you need to change how you access dimension types for lists?

Yes! we should change how model use `dim`. There are two spots currently in `VarGroup`, both of them is to determine the prior for the variables.
* We might need to ask user to input `lam` as a dictionary rather than a float number
* I am working on delegating `get_smoothing_gprior` to `Dimension` class itself in branch `feature/dim_smoothing_gprior` and hopefully will submit a PR soon.
* And for the use in the `__init__` of `VarGroup` it might be a little more complicated. We might forbidden the case where two categorical variable both specifying `lam`. Probably worth to address this in another PR.

Reply:
How would you deal with priors if one `type` is numerical and the other is categorical? How would you figure out the distance if both are numerical?